components/eam/src/chemistry/modal_aero/aero_model.F90

```diff
@@ -2083,76 +2099,75 @@ subroutine calc_schmidt_number( freepath, r_aer, temp,        & ! in
   end subroutine calc_schmidt_number
 
   !=============================================================================
-  subroutine qqcw2vmr(lchnk, vmr, mbar, ncol, im, pbuf)
-    use modal_aero_data, only : qqcw_get_field
+  subroutine qqcw2vmr(vmr, mbar, ncol, fldcw_all)
     !-----------------------------------------------------------------
     !	... Xfrom from mass to volume mixing ratio
+    ! C++ porting: this subroutine is similar to the subroutine mmr2vmr
+    ! in mozart/mo_mass_xforms.F90. Maybe can merge them
+    ! Note that the subroutine needs adv_mass from module chem_mod,
+    ! but the values are assigned from module mo_sim_dat
     !-----------------------------------------------------------------
 
     implicit none
 
     !-----------------------------------------------------------------
     !	... Dummy args
     !-----------------------------------------------------------------
-    integer, intent(in)     :: lchnk, ncol, im
-    real(r8), intent(in)    :: mbar(ncol,pver)
-    real(r8), intent(inout) :: vmr(ncol,pver,gas_pcnst)
-    type(physics_buffer_desc), pointer :: pbuf(:)
+    integer, intent(in)     :: ncol  ! indexes
+    real(r8), intent(in)    :: mbar(ncol,pver) ! mean wet atmospheric mass [g/mol]
+    real(r8), intent(in)    :: fldcw_all(ncol,pver,gas_pcnst) ! mass mixing ratio [kg/kg]
+    real(r8), intent(inout) :: vmr(ncol,pver,gas_pcnst) ! volume mixing ratios [mol/mol]
 
     !-----------------------------------------------------------------
     !	... Local variables
     !-----------------------------------------------------------------
-    integer :: k, m
-    real(r8), pointer :: fldcw(:,:)
-
-    do m=1,gas_pcnst
-       if( adv_mass(m) /= 0._r8 ) then
-          fldcw => qqcw_get_field(pbuf, m+im,lchnk,errorhandle=.true.)
-          if(associated(fldcw)) then
-             do k=1,pver
-                vmr(:ncol,k,m) = mbar(:ncol,k) * fldcw(:ncol,k) / adv_mass(m)
-             enddo
-          else
-             vmr(:,:,m) = 0.0_r8
-          endif
+    integer  :: kk, mm
+    real(r8) :: fldcw(ncol,pver) ! mass mixing ratio [kg/kg]
+
+    do mm=1,gas_pcnst
+       if( adv_mass(mm) /= 0._r8 ) then
+          fldcw = fldcw_all(:,:,mm)
+          do kk=1,pver
+             vmr(:ncol,kk,mm) = mbar(:ncol,kk) * fldcw(:ncol,kk) / adv_mass(mm)
+          enddo
```

Comment:
Do we need the `else` block here? We can also assign zeros to `vmr` in the beginning to avoid the `else` block:
```
vmr(:,:,:) = 0.0_r8
```
Can you try removing `ncol` from the argument list and define fldcw_all  and vmr as:
```
real(r8), intent(in)    :: fldcw_all(:,:,:) ! mass mixing ratio [kg/kg]
real(r8), intent(inout) :: vmr(:,:,:) ! volume mixing ratios [mol/mol]
```
Get rid of the `fldcw` local variable. In the do loop, you can do something like the following to avoid using `ncol` by using `fldcw_all`:
```
vmr(:,kk,mm) = mbar(:,kk) * fldcw_all(:,kk,mm) / adv_mass(mm)
```
This will make the I/O of these functions simpler. Also, If removing `ncol` takes time, please leave it as is.

Reply:
In other code I also saw similar issue when variables have dimension of `pver` and use `variable(:ncols)`. Is ncols always equal to pver? I am wondering if I want to remove ncols (or ncol) in the future.

Reply:
Maybe you meant to say `pcol` or `pcols` instead of `pver` (`pver` is the vertical dimension). `pcol` and `ncol` are related where `pcol` is always >=`ncol`. `ncol` can be smaller than `pcol` for some chunks.

This may cause issues sometimes if a variable is defined with a dimension of `pcol`. In that case, if `ncol < pcol`, the last elements of the array are undefined. If the undefined elements are used, they can result in a runtime error or NBFB answers.

Reply:
It should work fine in your case as you have defined `fldcw_all` using `ncol` in the `aero_model.F90` file.